src/vsr/client.zig

```diff
@@ -19,21 +19,130 @@ pub fn Client(comptime StateMachine_: type, comptime MessageBus: type) type {
         pub const StateMachine = StateMachine_;
 
         pub const Error = error{
+            BatchSizeZero,
+            TooMuchData,
             TooManyOutstandingRequests,
         };
 
-        pub const Request = struct {
-            pub const Callback = fn (
-                user_data: u128,
-                operation: StateMachine.Operation,
-                results: Error![]const u8,
-            ) void;
+        pub const Callback = fn (
             user_data: u128,
-            // Null iff operation=register.
-            callback: ?Callback,
+            operation: StateMachine.Operation,
+            results: Error![]const u8,
+        ) void;
+
+        pub const Request = struct {
+            // Null iif operation=register.
+            demux_list: ?DemuxList,
             message: *Message,
         };
 
+        pub const BatchLogical = struct {
+            operation: StateMachine.Operation,
+            size: u32,
+            batch_physical: union(enum) {
```

Comment:
Why not just always use `{ *Request, offset }`, and eagerly add the empty `Request` to the queue when `get_batch()` is called?

Reply:
I found it less error-prone not having to check for an eager/incomplete request, but rather keeping the check if the request queue was empty (eg `if (was_empty) self.send_request_for_the_first_time(message);`)